diff -r b49f6c4c060a core/init.lua
--- a/core/init.lua	Wed Jun 25 23:32:05 2008 -0400
+++ b/core/init.lua	Wed Jun 25 23:40:15 2008 -0400
@@ -35,7 +35,7 @@
     args = args..' --'..k
     if type(v) == 'string' then args = args..' "'..v..'"' end
   end
-  local p = io.popen(_HOME..'/core/lua_dialog.lua '..kind..args)
+  local p = io.popen('CocoaDialog '..kind..args)
   local out = p:read('*all')
   p:close()
   return out
diff -r b49f6c4c060a src/Makefile
--- a/src/Makefile	Tue Jun 24 18:43:59 2008 -0400
+++ b/src/Makefile	Wed Jun 25 14:44:26 2008 -0400
@@ -10,7 +10,7 @@ endif
 endif
 GTKFLAGS=$(shell pkg-config --cflags gtk+-2.0)
 GTKLIBS=$(shell pkg-config --libs gtk+-2.0 gthread-2.0)
-EXPORTLUASYMS=-rdynamic -Wl,--retain-symbols-file -Wl,lua.sym
+EXPORTLUASYMS=
 
 all: textadept
 
diff -r b49f6c4c060a src/textadept.c
--- a/src/textadept.c	Tue Jun 24 18:43:59 2008 -0400
+++ b/src/textadept.c	Wed Jun 25 14:44:26 2008 -0400
@@ -240,7 +240,7 @@ void new_scintilla_buffer(ScintillaObjec
   // Setup default styling and properties.
   SS(sci, SCI_STYLESETFONT, 32,
      reinterpret_cast<long>("!Bitstream Vera Sans Mono"));
-  SS(sci, SCI_STYLESETSIZE, 32, 8);
+  SS(sci, SCI_STYLESETSIZE, 32, 11);
   SS(sci, SCI_STYLESETFORE, 32, 0xAA | (0xAA << 8) | (0xAA << 16));
   SS(sci, SCI_STYLESETBACK, 32, 0x33 | (0x33 << 8) | (0x33 << 16));
   SS(sci, SCI_SETSTYLEBITS, 8, 0);
@@ -585,7 +585,7 @@ void set_default_editor_properties(Scint
   SS(sci, SCI_SETFOLDFLAGS, 16);
   SS(sci, SCI_SETMODEVENTMASK, SC_MOD_CHANGEFOLD);
 
-  SS(sci, SCI_SETMARGINWIDTHN, 0, 4 + 2 * // line number margin
+  SS(sci, SCI_SETMARGINWIDTHN, 0, 4 + 3 * // line number margin
      SS(sci, SCI_TEXTWIDTH, STYLE_LINENUMBER, reinterpret_cast<long>("9")));
 
   SS(sci, SCI_SETMARGINWIDTHN, 1, 0); // marker margin invisible
