--- a/lfs.c	2013-06-23 16:19:35.884341424 -0400
+++ b/lfs.c	2013-06-23 16:19:49.028452052 -0400
@@ -877,15 +877,6 @@
         {NULL, NULL},
 };
 
-#if LUA_VERSION_NUM > 501
-static void luaL_register (lua_State *L, const char *libname, const luaL_Reg *l)
-{
-        luaL_newlib (L, l);
-        lua_pushvalue (L, -1);
-        lua_setglobal (L, libname);
-}
-#endif
-
 int luaopen_lfs (lua_State *L) {
         dir_create_meta (L);
         lock_create_meta (L);
