diff -r 8a23edc91533 src/loadlib.c
--- a/src/loadlib.c	Mon Jan 12 18:57:53 2015 -0500
+++ b/src/loadlib.c	Mon Jan 12 23:50:08 2015 -0500
@@ -28,11 +28,11 @@
 ** variables that Lua check to set its paths.
 */
 #if !defined(LUA_PATH_VAR)
-#define LUA_PATH_VAR	"LUA_PATH"
+#define LUA_PATH_VAR	"TA_LUA_PATH"
 #endif
 
 #if !defined(LUA_CPATH_VAR)
-#define LUA_CPATH_VAR	"LUA_CPATH"
+#define LUA_CPATH_VAR	"TA_LUA_CPATH"
 #endif
 
 #define LUA_PATHSUFFIX		"_" LUA_VERSION_MAJOR "_" LUA_VERSION_MINOR
diff -r 8a23edc91533 src/luaconf.h
--- a/src/luaconf.h	Mon Jan 12 18:57:53 2015 -0500
+++ b/src/luaconf.h	Mon Jan 12 23:50:08 2015 -0500
@@ -166,24 +166,22 @@
 #define LUA_PATH_DEFAULT  \
 		LUA_LDIR"?.lua;"  LUA_LDIR"?\\init.lua;" \
 		LUA_CDIR"?.lua;"  LUA_CDIR"?\\init.lua;" \
-		LUA_SHRDIR"?.lua;" LUA_SHRDIR"?\\init.lua;" \
-		".\\?.lua;" ".\\?\\init.lua"
+		LUA_SHRDIR"?.lua;" LUA_SHRDIR"?\\init.lua;"
 #define LUA_CPATH_DEFAULT \
 		LUA_CDIR"?.dll;" \
 		LUA_CDIR"..\\lib\\lua\\" LUA_VDIR "\\?.dll;" \
-		LUA_CDIR"loadall.dll;" ".\\?.dll"
+		LUA_CDIR"loadall.dll;"
 
 #else			/* }{ */
 
-#define LUA_ROOT	"/usr/local/"
+#define LUA_ROOT	"/usr/"
 #define LUA_LDIR	LUA_ROOT "share/lua/" LUA_VDIR "/"
 #define LUA_CDIR	LUA_ROOT "lib/lua/" LUA_VDIR "/"
 #define LUA_PATH_DEFAULT  \
 		LUA_LDIR"?.lua;"  LUA_LDIR"?/init.lua;" \
-		LUA_CDIR"?.lua;"  LUA_CDIR"?/init.lua;" \
-		"./?.lua;" "./?/init.lua"
+		LUA_CDIR"?.lua;"  LUA_CDIR"?/init.lua;"
 #define LUA_CPATH_DEFAULT \
-		LUA_CDIR"?.so;" LUA_CDIR"loadall.so;" "./?.so"
+		LUA_CDIR"?.so;" LUA_CDIR"loadall.so;"
 #endif			/* } */
 
 
